jlm/llvm/opt/alias-analyses/RegionAwareModRefSummarizer.hpp

```diff
@@ -95,84 +98,151 @@ class RegionAwareModRefSummarizer final : public ModRefSummarizer
-   * The resulting SCCs and topological order will be stored in the `FunctionSCCs_` field.
+   * The resulting SCCs and topological order will be stored in the Context_ field.
    *
    * @param rvsdgModule the module for which a mod/ref summary is computed.
    */
   void
   CreateCallGraph(const rvsdg::RvsdgModule & rvsdgModule);
 
   /**
-   * Creates RegionSummaries and CallSummaries for the given function.
-   * These summaries will contain the memory locations utilized in each region / call,
-   * including in sub-regions and inside call targets.
-   *
-   * Assumes that all functions with lower sccIndex have already been annotated.
-   * If regions or calls may contain recursion, a flag is set, to signal that the summaries are not
-   * complete. By the time all functions in an SCC have been annotated, the field
-   * SCCMemoryNodesAffected_ will contain the complete set of memory locations affected by the SCC.
+   * For each SCC in the call graph, determines which allocas can be known to not be live
+   * when a function from the SCC is at the top of the call stack.
    */
   void
-  AnnotateFunction(const rvsdg::LambdaNode & lambda, size_t sccIndex);
+  FindAllocasDeadInSccs();
 
   /**
-   * Recursive call used to create summaries for the given region and all sub-regions.
-   * @param region the region to create summaries for.
-   * @param sccIndex the SCC index of the function being analyzed
+   * Creates a set containing all simple Allocas is the PointsToGraph.
+   * An Alloca is simple if it is only reachable from other simple Allocas,
+   * or from RegisterNodes, in the PointsToGraph.
    */
-  RegionSummary &
-  AnnotateRegion(rvsdg::Region & region, size_t sccIndex);
+  static util::HashSet<const PointsToGraph::MemoryNode *>
+  CreateSimpleAllocaSet(const PointsToGraph & pointsToGraph);
+
+  util::HashSet<const PointsToGraph::MemoryNode *>
+  GetSimpleAllocasReachableFromRegionArguments(const rvsdg::Region & region);
 
   /**
-   * Recursively annotates all subregions in the given structural node.
-   * Propagates memory locations to the parent region.
-   * @param structuralNode the structural node
-   * @param regionSummary the summary of the region containing the structural node
+   * Uses the call graph to determine if the given function can ever be involved
+   * in a recursive chain of function calls.
+   *
+   * @param lambda the function in question.
+   * @return true if it is possible for lambda to be involved in recursion, false otherwise
    */
-  void
-  AnnotateStructuralNode(
-      const rvsdg::StructuralNode & structuralNode,
-      RegionSummary & regionSummary);
+  bool
+  IsRecursionPossible(const rvsdg::LambdaNode & lambda);
 
-  void
-  AnnotateSimpleNode(const rvsdg::SimpleNode & simpleNode, RegionSummary & regionSummary);
+  /**
+   * Creates a set for each region that contains alloca definitions,
+   * where the alloca fits the requirements for being non-reentrant.
```

Comment:
What are the requirements for being non-reentrant? I have nowhere read these requirements yet.